x-pack/plugins/reporting/public/panel_actions/get_csv_panel_action.tsx

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import dateMath from '@elastic/datemath';
+import { i18n } from '@kbn/i18n';
+
+import { ContextMenuAction, ContextMenuActionsRegistryProvider } from 'ui/embeddable';
+import { PanelActionAPI } from 'ui/embeddable/context_menu_actions/types';
+import { kfetch } from 'ui/kfetch';
+import { toastNotifications } from 'ui/notify';
+
+const API_BASE_URL = '/api/reporting/v1/generate/immediate/csv/saved-object/';
+
+class GetCsvReportPanelAction extends ContextMenuAction {
+  constructor() {
+    super(
+      {
+        displayName: i18n.translate('xpack.reporting.dashboard.downloadCsvPanelTitle', {
+          defaultMessage: 'Download CSV',
+        }),
+        id: 'downloadCsvReport',
+        parentPanelId: 'mainMenu',
+      },
+      {
+        icon: 'document',
+      }
+    );
+  }
+
+  public async generateJobParams({ searchEmbeddable }: { searchEmbeddable: any }) {
+    const adapters = searchEmbeddable.getInspectorAdapters();
+    if (!adapters) {
+      return {};
+    }
+
+    if (adapters.requests.requests.length === 0) {
+      return {};
+    }
+
+    return searchEmbeddable.searchScope.searchSource.getSearchRequestBody();
+  }
+
+  public isVisible = (panelActionAPI: PanelActionAPI): boolean => {
+    const { embeddable } = panelActionAPI;
+
+    if (embeddable && embeddable.hasOwnProperty('savedSearch')) {
+      return true;
+    }
+
+    return false;
+  };
+
+  public onClick = async (panelActionAPI: PanelActionAPI) => {
+    const { embeddable } = panelActionAPI as any;
+    const {
+      timeRange: { from, to },
+    } = embeddable;
+
+    if (!embeddable) {
+      return;
+    }
+
+    const searchEmbeddable = embeddable;
+    const state = await this.generateJobParams({ searchEmbeddable });
```

Comment:
I think this needs to be wrapped in `_.pick` to whitelist fields that we care about. For example, we don't care about `highlight`, `size`, and `script_fields` (since they can't create script fields on-the-fly)

Reply:
Oh, this is just a comment of something we can figure out later, because I'm not sure what we need at the moment.

Reply:
Another thing (that doesn't need to be figured out right now) - this object doesn't seem to contain the columns. 

1. Create a saved search
2. Add it to a dashboard
3. Change the columns in the panel of the dashboard
4. We need to figure out how to get the columns to show, it's not in this `state` (but it could be)

Reply:
Sounds good!